services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java

```diff
@@ -79,49 +78,98 @@
 public final class ProtoReflectionService extends ServerReflectionGrpc.ServerReflectionImplBase
     implements InternalNotifyOnServerBuild {
 
-  private volatile ServerReflectionIndex serverReflectionIndex;
+  private ServerReflectionIndex serverReflectionIndex;
+
+  private Server server;
 
   private ProtoReflectionService() {}
 
-  public static BindableService getInstance() {
+  public static BindableService newInstance() {
     return new ProtoReflectionService();
   }
 
   /**
-   * Receives a reference to the server at build time.
+   * Do not use this method.
+   *
+   * @deprecated use {@link ProtoReflectionService#newInstance()} instead.
    */
+  @Deprecated
+  public static BindableService getInstance() {
+    return newInstance();
+  }
+
+  /** Receives a reference to the server at build time. */
   @Override
   public void notifyOnBuild(Server server) {
-    checkState(serverReflectionIndex == null);
-    serverReflectionIndex = new ServerReflectionIndex(checkNotNull(server, "server"));
+    this.server = checkNotNull(server);
+  }
+
+  /**
+   * Checks for updates to the server's mutable services and updates the index if any changes are
+   * detected. A change is any addition or removal in the set of file descriptors attached to the
+   * mutable services or a change in the service names.
+   */
+  private void updateIndexIfNecessary() {
```

Comment:
This is not thread-safe. `serverReflectionInfo()` may be called from multiple threads. Make it synchronized and put `@GuardedBy` on `serverReflectionIndex`?

You'd also need to return `serverReflectionIndex` from this method, to avoid grabbing the lock again in `serverReflectionInfo()`. Alternatively you could use a volatile, but meh.